reframe/frontend/loader.py

```diff
@@ -114,6 +118,17 @@ def prefix(self):
     def recurse(self):
         return self._recurse
 
+    def _set_defaults(self, test_registry):
+        for test in test_registry:
+            for name, val in self._external_vars.items():
+                if '.' in name:
+                    testname, varname = name.split('.', maxsplit=1)
+                else:
+                    testname, varname = test.__name__, name
+
+                if testname == test.__name__:
+                    setattr(test, varname, fields.make_convertible(val))
```

Comment:
Following the same argument as above, I think we might need to do something about this line. If a user sets a regular python attribute from the command line, this here would set it to a `_Convertible` type, which the user can't do much with, and probably get an error hard to debug.
So back to the metaclass (😭). What about moving the following lines into an independent function called `setvar` where `if name not in var_space: raise AttributeError(f'{name!} is not a variable')`
https://github.com/eth-cscs/reframe/blob/4b0cb19f72af8f4b9f954133a54240011bd510e8/reframe/core/meta.py#L422-L431
Then, the `__setattr__` method in the metaclass can just call the `setvar` method in that first `try/except` block and the rest is the same. That would allow us to be 100% sure that the users are setting variables and variables only from the command line.

Reply:
One little problem I'm facing here is this. When calling `setvar()` from `__setattr__()` in the metaclass, I'd like to differentiate the error thrown when a variable is not defined and the syntax error `cannot override variable descriptor`. The obvious way would be to raise an `AttributeError`, but I'm reluctant for two reasons: (a) the error is not an attribute error (b) I don't want another try/catch in the loader to convert the error to something friendlier. What I've gone for is to define a `VariableError` that derives from `ReframeSyntaxError`. Currently, I'm throwing the `VariableError` only here, and everything works fine, but would it make sense to throw it in other locations, too? However, if we go for "throw a VariableError for any variable error," then the `cannot override variable descriptor` isn't it a variable error as well? So we're back to the start that we can't differentiate the errors raised in `setvars()`. To avoid this I'm stating that `VariableError` is raised only when a reference to a variable fails, similarly to `AttributeError`. Are we happy with this?

Reply:
Hmm, I just hit a more fundamental problem. Check this:

```
./bin/reframe -c unittests/resources/checks_unlisted/externalvars.py -r -S external_x.foo=3 -S external_y.foo=2 -S foolist=3,4
```

`foolist` is defined only in the `external_y`, so if we choose to throw a hard error for trying to set `external_x`, then we practically abort loading the the whole file. I would argue, that we simply skip inexistent variables and not set them. I'm not sure if we even want to issue a warning, because it could lead to a flood of warnings without a big reason.